typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,717 @@
+
+Using all this new syntax, we can define a typing rule for abstractions now:
+
+$$
+\begin{array}{c}
+\Gamma, x:T_1 \vdash t : T_2 \\
+\hline
+\Gamma \vdash \lambda x:T_1.\; t : T_1 \to T_2
+\end{array}\quad\text{(Abs)}
+$$
+
+This rule simply says that if -- given the current context plus the type of the abstraction variable $T_1$ -- the type of the body turns out to be $T_2$, then the whole abstraction in the current context has type $T_1 \to T_2.$
+Pretty straightforward, but the notation may seem a bit arcane.
+Spend a minute to understand it, we'll be using it more down the line.
+
+For individual variables, the typing rule would then be rather trivial:
+
+$$
+\begin{array}{c}
+x : T \in \Gamma \\
+\hline
+\Gamma \vdash x : T
+\end{array}\quad\text{(Var)}
+$$
+
+For applications, it's a bit more involved, but hopefully nothing too surprising:
+
+$$
+\begin{array}{c}
+\Gamma \vdash t_1 : T_1 \to T_2;\quad \Gamma\vdash t_2 : T_1 \\
+\hline
+\Gamma \vdash t_1\; t_2 : T_2
+\end{array}\quad\text{(App)}
+$$
+
+Try to interpret this one yourself, then check if you're correct.
+
+<p><details>
+<summary>Answer</summary>
+
+This reads, basically, "for an application expression, if the left-hand side (function) has type $T_1\to T_2,$ and the right-hand-side (argument) has the type $T_1$, then the result of the application will have type $T_2.$
+
+***
+
+</details></p>
+
+To clear up a technicality, we also need to have the context in our $\text{(If)}$ rule, since typings of terms might depend on the context, too:
+
+$$
+\begin{array}{cc}
+\begin{array}{c}
+\Gamma\vdash t_1 : Bool;\quad \Gamma\vdash t_2 : T;\quad \Gamma\vdash t_3 : T \\
+\hline
+\Gamma\vdash\mathbf{if}\;t_1\;\mathbf{then}\;t_2\;\mathbf{else}\;t_3 : T
+\end{array}
+& \text{(If)}
+\end{array}
+$$
+
+### Example
+
+The practical application of these rules is we can use those to assign types to terms and thus ensure the program doesn't get stuck.
+
+To do that, we build typing trees, which just chain typing rules in a tree-like structure.
+
+As a simple example, let us show that term $(\lambda x: Bool.\; x)\;true$ has type $Bool$ in some arbitrary context $\Gamma$.
+
+To build the typing tree, we start with the term in question and find the typing rule where the consequent matches it best.
+In this case, this is the $\text{(App)}$ rule:
+
+$$
+\begin{array}{c}
+\Gamma \vdash t_1 : T_1 \to T_2, t_2 : T_1 \\
+\hline
+\Gamma \vdash t_1\; t_2 : T_2
+\end{array} \quad\text{(App)}
+$$
+
+Here we annotate each typing rule with a name for clarity.
+
+We then instantiate all placeholders we can:
+
+$$
+\begin{array}{c}
+\Gamma \vdash (\lambda x:Bool.\; x) : T_1 \to T_2;\quad \Gamma \vdash true : T_1 \\
+\hline
+\Gamma \vdash (\lambda x:Bool.\; x)\; true : T_2
+\end{array} \quad\text{(App)}
+$$
+
+For each condition, we do the same in turn:
```

Comment:
I'm not sure typing trees work this way. What you described looks like an informally-specified type inference algorithm, but as far as typing trees themselves are concerned, you're simply supposed to magically know the right answers.

So I agree with the typing tree you give at the end of this section, whereas the intermediate result is not a valid typing tree because it's not valid to selectively instantiate parts of it. It looks like an intermediate data structure that could be used by an inference algorithm.

I guess my problem is only with presentation. It is easy to read this section and end up with the misconception that there's a canonical method to construct typing trees. In reality, there might be more than one algorithm for a given type system, or the type system could be undecidable. To avoid discussing any algorithms, you can assume the existence of an oracle that just magically tells you the type of a term, and then building the typing tree merely validates that the oracle didn't lie to you. (In a compiler, we have the type inference engine instead of the oracle)